runelite-client/src/main/java/net/runelite/client/plugins/chatcommands/ChatCommandsOverlay.java

```diff
@@ -0,0 +1,154 @@
+ *	and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.chatcommands;
+
+import java.awt.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.inject.Inject;
+
+import net.runelite.api.Client;
+import net.runelite.api.Point;
+import net.runelite.api.widgets.ComponentID;
+import net.runelite.api.widgets.Widget;
+import net.runelite.client.ui.overlay.*;
+import net.runelite.client.ui.overlay.tooltip.Tooltip;
+import net.runelite.client.ui.overlay.tooltip.TooltipManager;
+
+public class ChatCommandsOverlay extends Overlay
+{
+	private final Pattern PETS_MATCHER = Pattern.compile("<img=([0-9]+)>", Pattern.MULTILINE);
+
+	private final Client client;
+	private final ChatCommandsPlugin plugin;
+	private final ChatCommandsConfig config;
+
+	@Inject
+	private TooltipManager tooltipManager;
+
+	@Inject
+	private ChatCommandsOverlay(Client client, ChatCommandsPlugin plugin, ChatCommandsConfig config)
+	{
+		super(plugin);
+		setPosition(OverlayPosition.DYNAMIC);
+		setLayer(OverlayLayer.ABOVE_WIDGETS);
+		setPriority(Overlay.PRIORITY_MED);
+		this.client = client;
+		this.plugin = plugin;
+		this.config = config;
+	}
+
+	private void petsOverlay(Graphics2D graphics)
+	{
+		// Still loading icon data
+		if (plugin.petsIconIdx == -1)
+		{
+			return;
+		}
+		// Get the chat box lines
+		Widget messageLines = client.getWidget(ComponentID.CHATBOX_MESSAGE_LINES);
+		if (messageLines == null)
+		{
+			return;
+		}
+		Widget[] lines = messageLines.getChildren();
+		if (lines == null)
+		{
+			return;
+		}
+		for (Widget line : lines)
+		{
+			// ensure visible, text is not null, and matching pets text
+			if (messageLines.getBounds().contains(line.getBounds())
+					&& line.getText() != null
+					&& line.getText().matches(".*?Pets: \\([0-9]+\\) <img=[0-9]+>.*"))
+			{
+				// Get the bounds
+				Rectangle bounds = line.getBounds();
+				// Get mouse position
+				Point mousePos = client.getMouseCanvasPosition();
+				// If mouse in message bounds
+				if (bounds.contains(mousePos.getX(), mousePos.getY()))
```

Comment:
Instead of running bounds checking every frame a pets chat message exists in a player's chat history (visible or otherwise), this could also be gated on checking that the hovered widget is a chat message and checking the hovered message's text.

Reply:
Did a bunch of measuring: 

Limitation of the menu method is you can't hover your own pets which would make for a weird UX. 
However it obviously is faster. 

Compared against my optimized is visible checking it is: 
## Profiling
*72 samples averaged out continuously over 2 ticks at 60fps. All samples with a full chatbox and multiple pets commands*

### Only Others' Pets (onMenuEntryAdded)
Idle: ~0.5ms
Active: ~5.3ms

### Optimized Visibility Checking
*Includes manual pet icon id parsing w/o regex*
Idle: ~1ms
Idle (mouse in chat box): ~8.5ms
Active: ~9.2ms

Delta from empty to full chat box is ~2ms

Delta in JVM instructions from the original bound check and the new visible checking is ~147 instructions down to ~36